### ISSUE TITLE
Peak annotations in spectra

### ISSUE THREAD

**Issue by lgatto** [lgatto/MSnbase#456]
This issue will be used to discuss models and implementation of having peak annotations in spectra. It refers, among others to discussion at the recent metabolomics workshop and works by @meowcat in his [`MSnbase` fork](https://github.com/meowcat/MSnbase).

Cc: @jorainer @sgibb @Treutler @michaelwitting

I will try to summarise a few points based on today's dev call with @jorainer, but it might be a good thing to consider another call dedicated to this issue with all interested parties in the coming weeks.

PS: this also considers the spectum-level `metadata`.

**Comment by meowcat** [comment]
I am looking forward to a devcall! Herein a few considerations that have come up in email and I thought about some more and I ended up with a new suggestion (It happens that I start writing long posts when I think about issues, so I apologize for the wall of text...) I write this not to shut alternative ideas down, but I think it's probably better understandable if you can read it here rather than "live" at the devcall.

# New suggestion:
* Define an `AnnotatedSpectrum` class with @metadata and @peakAnnotations that derives from `Spectrum2`, but can be constructed from either `Spectrum1` or `Spectrum2` objects, and represent either spectrum
* The `` `[[` `` operator on the `Spectra` object could return an `AnnotatedSpectrum` . This would allow you to write backends without caring about the annotations (however those should be able to have backends too, e.g. a database!)
* (If users want to save memory on MS1 spectra, one could set a flag in the `Spectra` object whether to return `AnnotatedSpectrum` or plain `Spectrum1/2`.)

This would allow the user to handle single spectra as an entity, which is extremely helpful in practice!


# Long train of thought:
* **If performance in the constructor is the main issue**, it could be remedied with initializing the fields to NULL (and having class union listORnull, dataframeORnull there), and initializing them on first access or optionally on construction, since in reality many people will never use these two slots for most processing tasks. (There is really no justification for making the package much slower and bulkier for most users just because a few hackers want their slots in the `Spectrum` class. I understand that.)
* We have a ton of code (e.g. in-house scripts at Eawag) that works with a single Spectrum, and with single RmbSpectrum2 (which is a Frankenstein version of what I propose to implement for `peakAnnotations` and `metadata`).
* I understand that there can be length-one `Spectra` objects. But (`specs` being a `Spectra` object) would that mean that we have to use `specs[[1]]`, `mz(specs)[[1]]` for every single call? Could there be a less clunky solution to that? - If we end up writing a Spectrum wrapper around Spectra, I think we are back at step one...
* Also, in this case the annotation is lost when I do `sp <- spectra[[1]]` and I have to keep it separate (but not when using `spectra[1]`). This is similar to a solution I was using **before** I added the `@metadata` (or, how it is called there, `@info`) to `RmbSpectrum2`. Having the data **in** the `Spectrum` was an incredible leap forward for data handling; think doing (`specs`  being a `Spectra` object of length n)
  `processedSpecs <- lapply(specs, processingFunction)` 
instead of 
 ```
results <- lapply(seq_along(specs), function(sp) otherFunction(specs[[i]], peakAnnotations(specs)[[i]], metadata(specs)[[i]]
processedSpecs <- Spectra(results$specs, results$peakAnnotations, results$metadata)
```
or maybe
```
specs <- lapply(specs, function(sp) otherFunction(sp[[1]], 
  peakAnnotations(sp)[[1]], metadata(sp)[[1]] )`
# I understand that I can pass sp directly to otherFunction, but I will
# then essentially be doing the same thing *inside* the function!
```
This is (conceptually) what a lot of RMassBank code and my inhouse scripts looked like **before** I added slots to `Spectrum2`, which made everything much nicer. Some of it still does in the Bioconductor version (less so in the `s4power` branch). 
* It could be argued that the solution should be to vectorize my functions, and this is possible in some cases, but in many cases a `Spectrum` as an entity is just very convenient to work with! For `peakAnnotations` even more so than for the `metadata`. There are not so many applications where I would *not* end up looping over all `spectra, peakAnnotations, metadata` in parallel.
* Also note the behavior of the proposed `setData` and `as.data.frame` (or if you want to avoid trouble, rename it to `getData`).

## Not brought up yet:
* Spectrum `metadata` can be expected to be quite ragged, and in my use case (but not necessarily for everyone) hierarchical - so if at all storing it in a dataframe, I don't think storing it in columns per attribute would be very intuitive. In a database setting, I would certainly store those in a `TABLE metadata (spectrum_id INT, metadata_id INT, value ???)` , `TABLE metadata_keys (metadata_id INT, name VARCHAR,` and possibly `metadata_parent INT)` format.
Think library spectra loaded from different sources, that have very different metadata annotated.
* Another point: are backends for `Spectra@metadata` planned as for now? 

## Connected topic, and then the suggestion
* I have at the hackathon briefly mentioned (but not elaborated) that I am not sure that the distinction of `Spectrum1` and `Spectrum2` is essential. I feel this can probably not be changed at this point... I guess one reason is also to save memory for MS1 scans?
* If you look at the current behavior of `Spectra`, it seems to me it is essentially behaving as if everything was a `Spectrum2` and filling everything from `Spectrum1` with `NA`. 
* (Also conceptually, there are things that only halfway fit - low-energy/high-energy MSe scans for example, DIA with large windows where the "precursor" is really not a mass but a range, ...)
* From this came my above-stated suggestion to make an `AnnotatedSpectrum` class which can represent both `Spectrum1` and `Spectrum2`, since `Spectrum1` is a subset of `Spectrum2` feature-wise.

**Comment by meowcat** [comment]
Two more advantages:
* a first implementation in Spectra is straightforward, putting the `peakAnnotation` dataframe in a specialized column in `Spectra@metadata`, and worrying about the raggedness issue later
* Implementation timelines are uncoupled

If you want, i can prepare some code for this.

**Comment by lgatto** [comment]
Thank you @meowcat. 

In summary, I am in favour of adding *fields* for general metadata (note however that this one exists already in `Spectra`) and support for peak annotations. There is however some confusion about current infrastructure, future directions of `MSnbase` and how spectra data is encoded.

The main point is that conceptually, `Spectrum` and `Spectra` are the same thing. The fundamental difference is the concept of backend, that doesn't exists for a `Spectrum`, but it central in `Spectra`. Backends are the result from serious limitations in terms of speed and memory that stem from the in memory S4 implementation of `Spectrum`. In other words, there are indeed issues in terms of speed and memory, but these are not those that you are alluding to above. I will describe this in more details during the call. 
To already address your point above that wonders about `Spectra[[1]]` vs `Spectrum`,  the answer is that one would always use `Spectra`, irrespective of `length(Spectra)` - `Spectra` are by default vectorised. 

Regarding peak annotations, I don't think a `data.frame` would fit here (I don't think you mention this above, but this is what has been reported to me). There are many open issues regarding peak annotations (not the least of them is matching annotations to peaks for centroided and profile mode data), and it isn't clear to me yet what the best approach is to implement such annotations. However, my first inkling would be to define a `PeaksAnnotation` object, that itself can be exist as a list of `PeakAnnotations` with vectorised operations. Again, this will I think make much more sense during the call, when the reasons why all our efforts are aimed at the `Spectra` class will become clear.

Looking forward to talking to you!

**Comment by meowcat** [comment]
Some fair points. I have to understand the scope and usage of `Spectra` better; as of now (`Spectra` from MSnbase 2.9.2) there is considerable difference in how they work (purely in terms of usage, not logic) for the end user for some simple things I tried out; I mean that mostly for functionality that goes beyond using the provided methods, and concerning the user-turned-scripter/coder population. I will be happy to learn (and also re-educate my coworkers who I hounded to use the `Spectrum` classes [instead of their own constructs] in the last few years.)

WRT the data.frame implementation of peak level annotations, it has served me well for a while for my (admittedly quite specific) purpose. The main drawback to me is a 1:1 mapping between peak and annotations, such that e.g. multiple formulas for one peak are not possible (I replicated peaks for this, which is unsatisfying.) I guess this also relates to your issue with centroided vs profile peaks. 
* To expand on your first ideas, consider that some annotation columns are linked to each other; example is "predicted formula" and "mass error (of the predicted formula)" whereas others are conceptually disconnected (e.g. "signal-to-noise ratio".
* Another first inkling I have to address centroid vs profile mode and 1:n annotations would be 
  * [peakId, mz, columns] per annotation,
  * where peakId matches to an individual peak in the spectrum, (then mz is left empty), 
  * or mz can define a custom position not necessarily contained in the original spectrum (then peakId is left empty.)
Just throwing thoughts out there.
